tests/channel_int/channel_int.c

```diff
@@ -0,0 +1,513 @@
+}
+
+static void test_free_channel_data2(void* data)
+{
+    ASSERT_ARE_EQUAL(void_ptr, test_data2, data);
+}
+
+static THANDLE(THREADPOOL) setup_threadpool()
+{
+    EXECUTION_ENGINE_HANDLE execution_engine = execution_engine_create(NULL);
+    ASSERT_IS_NOT_NULL(execution_engine);
+    THANDLE(THREADPOOL) threadpool = threadpool_create(execution_engine);
+    ASSERT_IS_NOT_NULL(threadpool);
+    ASSERT_ARE_EQUAL(int, 0, threadpool_open(threadpool));
+    return threadpool;
+}
+
+static void pull_callback_order_checker(void* context, CHANNEL_CALLBACK_RESULT result, THANDLE(RC_PTR) data)
+{
+    ASSERT_IS_NOT_NULL(context);
+    ASSERT_ARE_EQUAL(CHANNEL_CALLBACK_RESULT, CHANNEL_CALLBACK_RESULT_OK, result);
+    ASSERT_IS_NOT_NULL(data);
+
+    //assert that n-th pull is matched with n-th push
+     ASSERT_ARE_EQUAL(int32_t, *((int32_t*)context), (int32_t)(int64_t)((void*)data->ptr)); //casts needed to suppress pointer truncation warning
+
+    free(context);
+}
+
+static int pull_data(void* context)
+{
+    THANDLE(CHANNEL) channel = context;
+    for (int32_t i = 1; i <= CHANNEL_ORDER_TEST_COUNT; i++)
+    {
+        int32_t* pull_id = malloc(sizeof(int32_t));
+        *pull_id = i;
+        THANDLE(ASYNC_OP) async_op = NULL;
+        CHANNEL_RESULT pull_result = channel_pull(channel, pull_callback_order_checker, pull_id, &async_op);
+
+        ASSERT_ARE_EQUAL(CHANNEL_RESULT, CHANNEL_RESULT_OK, pull_result);
+        ASSERT_IS_NOT_NULL(async_op);
+
+        THANDLE_ASSIGN(ASYNC_OP)(&async_op, NULL);
+    }
+    return 0;
+}
+
+static void dummy_free_func(void* ptr)
+{
+    (void)ptr;
+}
+
+static void push_callback_order_checker(void* context, CHANNEL_CALLBACK_RESULT result)
+{
+    ASSERT_IS_NOT_NULL(context);
+    ASSERT_ARE_EQUAL(CHANNEL_CALLBACK_RESULT, CHANNEL_CALLBACK_RESULT_OK, result);
+
+    interlocked_increment(&g_push_callback_count);
+    wake_by_address_single(&g_push_callback_count);
+}
+
+static int push_data(void* context)
+{
+    THANDLE(CHANNEL) channel = context;
+    for (int32_t i = 1; i <= CHANNEL_ORDER_TEST_COUNT; i++)
+    {
+        THANDLE(RC_PTR) data = rc_ptr_create_with_move_pointer((void*)(int64_t)i, dummy_free_func);
+        THANDLE(ASYNC_OP) async_op = NULL;
+        CHANNEL_RESULT push_result = channel_push(channel, data, push_callback_order_checker, (void*)(int64_t)i, &async_op);
+
+        ASSERT_ARE_EQUAL(CHANNEL_RESULT, CHANNEL_RESULT_OK, push_result);
+        ASSERT_IS_NOT_NULL(async_op);
+
+        THANDLE_ASSIGN(RC_PTR)(&data, NULL);
+        THANDLE_ASSIGN(ASYNC_OP)(&async_op, NULL);
+    }
+    return 0;
+}
+
+BEGIN_TEST_SUITE(TEST_SUITE_NAME_FROM_CMAKE)
+
+TEST_SUITE_INITIALIZE(suite_init)
+{
+    g_execution_engine = execution_engine_create(NULL);
+    ASSERT_IS_NOT_NULL(g_execution_engine);
+    THANDLE(THREADPOOL) temp = threadpool_create(g_execution_engine);
+    ASSERT_IS_NOT_NULL(temp);
+    THANDLE_INITIALIZE_MOVE(THREADPOOL)(&g.g_threadpool, &temp);
```

Comment:
>g_threadpool [](http://example.com/codeflow?start=43&length=12)

all these tests should run with the default config of the treadpool and with a config of "1" thread only in the threadpool.